icu4c/source/common/unicode/localpointer.h

```diff
@@ -548,46 +548,52 @@ class LocalArray : public LocalPointerBase<T> {
  * @stable ICU 4.4
  */
 #define U_DEFINE_LOCAL_OPEN_POINTER(LocalPointerClassName, Type, closeFunction) \
-    class LocalPointerClassName : public LocalPointerBase<Type> { \
-    public: \
-        using LocalPointerBase<Type>::operator*; \
-        using LocalPointerBase<Type>::operator->; \
-        explicit LocalPointerClassName(Type *p=nullptr) : LocalPointerBase<Type>(p) {} \
-        LocalPointerClassName(LocalPointerClassName &&src) noexcept \
-                : LocalPointerBase<Type>(src.ptr) { \
-            src.ptr=nullptr; \
-        } \
-        /* TODO: Be agnostic of the deleter function signature from the user-provided std::unique_ptr? */ \
-        explicit LocalPointerClassName(std::unique_ptr<Type, decltype(&closeFunction)> &&p) \
-                : LocalPointerBase<Type>(p.release()) {} \
-        ~LocalPointerClassName() { if (ptr != nullptr) { closeFunction(ptr); } } \
-        LocalPointerClassName &operator=(LocalPointerClassName &&src) noexcept { \
-            if (ptr != nullptr) { closeFunction(ptr); } \
-            LocalPointerBase<Type>::ptr=src.ptr; \
-            src.ptr=nullptr; \
-            return *this; \
-        } \
-        /* TODO: Be agnostic of the deleter function signature from the user-provided std::unique_ptr? */ \
-        LocalPointerClassName &operator=(std::unique_ptr<Type, decltype(&closeFunction)> &&p) { \
-            adoptInstead(p.release()); \
-            return *this; \
-        } \
-        void swap(LocalPointerClassName &other) noexcept { \
-            Type *temp=LocalPointerBase<Type>::ptr; \
-            LocalPointerBase<Type>::ptr=other.ptr; \
-            other.ptr=temp; \
-        } \
-        friend inline void swap(LocalPointerClassName &p1, LocalPointerClassName &p2) noexcept { \
-            p1.swap(p2); \
-        } \
-        void adoptInstead(Type *p) { \
-            if (ptr != nullptr) { closeFunction(ptr); } \
-            ptr=p; \
-        } \
-        operator std::unique_ptr<Type, decltype(&closeFunction)> () && { \
-            return std::unique_ptr<Type, decltype(&closeFunction)>(LocalPointerBase<Type>::orphan(), closeFunction); \
-        } \
+    using LocalPointerClassName = LocalOpenPointer<Type, closeFunction>
+
+#ifndef U_IN_DOXYGEN
+template <typename Type, auto closeFunction>
+class LocalOpenPointer : public LocalPointerBase<Type> {
```

Comment:
problem: This is now a newly, publicly defined class. As such, it should be `@internal` -- or `@draft ICU 75` with a proposal.

Reply:
I didn't appreciate the doxygen guard, but it's still confusing to someone reading the header file.
It would be good to add an `@internal` comment.

Reply:
Maybe more interestingly, if you are in an IDE and look at the class, it would help to have a doc string that says it's internal.

Reply:
Does such a thing as an IDE that expands the macro and then resolves the generated `using`-statement and then displays the docstring for the underlying class template really exist? (If it does, then sure, then it certainly would be meaningful to add an `@internal` annotation here.)

Reply:
To ensure that anyone poking around in the header file will understand that this class template is internal, I propose wrapping it in a nested namespace named `internal` as the most effective solution, that should be impossible to miss in any context: PR #2815

Reply:
ok that looks acceptable tnx

Reply:
> (If it does, then sure, then it certainly would be meaningful to add an `@internal` annotation here.)

Now when Andy has shown that it really does: PR #2816